BEHST update

davidechicco opened the pull request:
* [X] I have read the [guidelines for bioconda recipes](https://bioconda.github.io/guidelines.html).
* [ ] This PR adds a new recipe.
* [X] This PR updates an existing recipe.
* [ ] This PR does something else (explain below).

I edited my Bitbucket package and my recipe as suggested, by removing my data files from the Bitbucket package and adding a download script which downloads the data from the internet. My recipe passed the tests on Travis CI

bgruening commented on recipes/BEHST/meta.yaml:
```diff
@@ -0,0 +1,46 @@
+    - r-base
+    - curl
+    - r-gprofiler
+    - gnu-wget
```
This is only a build dependency.

davidechicco replied:
gnu-wget? Yes, I can remove it.

bgruening commented on recipes/BEHST/meta.yaml:
```diff
@@ -0,0 +1,46 @@
+    - bedtools
+    - pybedtools
+    - r-base
+    - curl
```
build dependency? or really needed during runtime?

davidechicco replied:
Probably not needed; I'll try to remove them.

bgruening commented on recipes/BEHST/meta.yaml:
```diff
@@ -0,0 +1,46 @@
+    - gnu-wget
+    
+test:
+  files:
```
you haven't really included a test ...

You should add something like `project.sh --help`

davidechicco replied:
What do you mean? The test is the execution of `project.sh`

bgruening commented on recipes/BEHST/meta.yaml:
```diff
@@ -0,0 +1,46 @@
+  sha256: cf4cc7179e6311f72d350c4fd408930b367b568f4ca77469474c4e5044ffd015
+
+build:
+  number: 1
```
build should start with 0

davidechicco replied:
Okay but why?

daler replied:
https://conda.io/docs/building/meta-yaml.html#build-number-and-string: "The build number should be incremented for new builds of the same version. The number defaults to zero."

If you re-build version 0.4 with some fixes to the recipe, then you would increase the build number.

bgruening commented on recipes/BEHST/build.sh:
```diff
@@ -0,0 +1,13 @@
+set -o nounset -o pipefail -o errexit
+set -o xtrace
+
+PREFIX=.
```
Not needed.

davidechicco replied:
Okay

bgruening commented on recipes/BEHST/build.sh:
```diff
@@ -0,0 +1,13 @@
+cd $PREFIX/bin/
+
+
+./download_behst_data.sh
```
I'm not really sure how your installation work. But in the end your binaries needs to end up in $PREFIX/bin. You are not allowed to change $PREFIX :(

davidechicco replied:
I really cannot understand this $PREFIX story...

How should I edit my `build.sh` and `test.sh` files? Thanks

bgruening replied:
@davidechicco we have some nice documentation here: https://bioconda.github.io/contribute-a-recipe.html
Maybe this helps you? It also includes the md5 sha256 things.

davidechicco replied:
@bgruening Thanks but I cannot find any information about the `$PREFIX` issue. Anyway, I edited my `build.sh` and `test.sh` scripts and now their 1st line is `cd $PREFIX/bin/`
I hope this satisfies the requirements.

daler replied:
From https://conda.io/docs/building/recipe.html#conda-recipe-files-overview, the building takes place in the source directory. `$PREFIX` is the location defined by conda-build to which the build script should install executables (https://conda.io/docs/building/environment-vars.html#environment-variables-set-during-the-build-process). In simple cases, this is:

```bash
cp executable $PREFIX/bin
```

davidechicco replied:
@daler I tried to put the executable files in `$PREFIX/bin` but my `$PREFIX` folder (which is `/opt/conda/conda-bld/behst_1498494985724`) does not contain any `/bin/` folder.

I switched to `cd $SRC_DIR/bin` and everything works fine on Travis. Is this setting correct? Thanks

daler replied:
Sorry I should update the simple case to:

```bash
mkdir -p $PREFIX/bin
cp executable $PREFIX/bin
```

You might want to take a look at some other recipes to see how things are done.

davidechicco replied:
Okay, I added the followed commands to `build.sh`:
`mkdir -p $PREFIX/bin`
`cp $SRC_DIR/bin/* $PREFIX/bin`
`cd $PREFIX/bin`

and it worked on Travis. Is this version alright?

daler commented:
If you're fine with making changes to BEHST, the following changes will make recipe building much easier and will make your code more portable even when not used with conda:

- `project.sh` should print help and should test for presence of the helper scripts. Maybe a `--test` argument to trigger this check. Then the conda recipe's test becomes `project.sh --test` since all we want to do in a conda recipe test is to make sure executables are on the path
- `project.sh` should not have hard-coded filenames, they should be provided as commandline args. At the very least, the data directory should be specified from the command line. This is what's making the recipe so difficult.
- the new download data script should correspondingly accept a commandline arg for where to download data.

#5053 is an attempt (still not working) at this recipe. The above upstream changes will make things much easier to package.

davidechicco commented:
@daler Hi, I addressed most of your suggestions in the new version of my package. Now `project.sh` accepts the data file names as arguments, and in case the argument is `--test`, it checks the existence of the executable files. I edited my bioconda recipe `test.sh` accordingly. 
I do not agree with your suggestion on the dowload file. I think it should download the default data files in the `/data/` folder as it is right now, without letting the user choose. The user can put his/her own data file anywhere he/she wants and use them as input to `project.sh`, but I don't think he/she should be able to decide where to put my default data files.

The new tests on Travis CI passed, so I really hope you can finally create the BEHST recipe on bioconda.

Thanks

daler commented:
Great. I see this PR and also #4957. This one is newer but the other has been updated more recently. Which one should I be looking at?

davidechicco commented:
The newer one

daler commented:
So #4957 (older one but recently updated) or #5047 (newer one)?

Also, about the default data location: I understand that you want to control where it goes, but as an example of why you might want users to control where it goes, I keep my conda installation in my home directory on our cluster. We are not given a large quota in that directory. I can't use your package if it downloads data there because it would fill up my quota, but I could use it if I could put the data in location where I have tons of space.

davidechicco commented:
> So #4957 (older one but recently updated) or #5047 (newer one)?

I don't know... the last version of my recipe is running now on [Travis CI](https://travis-ci.org/davidechicco/bioconda-recipes/builds/247586929) and its code is [here](https://github.com/davidechicco/bioconda-recipes/tree/master/recipes/BEHST)   

For the data folder, I understand your concert but I need users to be able to use some DEFAULT_* arguments regarding the input data files for `project.sh`
To be default, their file paths must be known. If users want to use their dataset files is fine, but letting them set the default input files is too complicated. Please notice most of the users of BEHST will be biologists, so people without a specific knowledge of programming.

Thanks

daler commented:
OK. It looks like #4957 is the PR you opened. I will make comments over there.

daler commented:
Closing in favor #4957

davidechicco commented:
Hi @daler 
sorry to disturb you again. I decided to replace my `behst.py` script with the `behst` script (without extension). I tried to download the last bioconda project, do my edits, and the commit and push the code, but it did not work. I just edited the `run_test.py` script, by replacing `behst.py` with `behst`.

Perhaps I do not have the right to perform a push?
Here's the error I get:

> Total 5 (delta 4), reused 0 (delta 0)
> remote: Resolving deltas: 100% (4/4), completed with 4 local objects.
> remote: error: GH006: Protected branch update failed for refs/heads/master.
> remote: error: Required status check "continuous-integration/travis-ci" is expected.
> To https://github.com/bioconda/bioconda-recipes.git
>  ! [remote rejected] master -> master (protected branch hook declined)
> error: failed to push some refs to 'https://github.com/bioconda/bioconda-recipes.git'

Thanks!

daler commented:
You need to create a branch, and then push that, then open a PR on that branch:

```bash
git checkout -b my-new-branch-name
git push origin my-new-branch-name
```

Then on the github site, when you visit the bioconda repo, you will get a message saying you just pushed a branch and it offers to open a pull request.

The reason is that the master branch is protected, and recipes can only make it in after they have passed all tests.

davidechicco commented:
Hi @daler 
Thanks for the reply. I created a new branch (`behst4`), edited my code, committed and tried to push with the `git push origin behst4`  command, but I'm getting this error:

```
[davide@opennet-33-58 behst]$ git push origin behst4;
Username for 'https://github.com': davidechicco
Password for 'https://davidechicco@github.com': 
To https://github.com/bioconda/bioconda-recipes.git
 ! [rejected]        behst4 -> behst4 (fetch first)
error: failed to push some refs to 'https://github.com/bioconda/bioconda-recipes.git'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first merge the remote changes (e.g.,
hint: 'git pull') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```
Any idea on how to solve this issue? Thanks

davidechicco commented:
Please ignore the previous message. I re-cloned the repository and created a new branch, `behst5`, and apparently it works now. I made a new 

> https://github.com/bioconda/bioconda-recipes/pull/5527

.

davidechicco commented:
Hi @daler 
I made other edits to my package and my recipe, and tried to linked it to the current version (1.4) of my package. I did my edits on the `behst7` [branch](https://github.com/bioconda/bioconda-recipes/tree/behst7/recipes/behst) and made a pull request. I thought it work but it did not. The current master bioconda project still links to the 0.9 version of BEHST.

The `behst7` branch webpage says also "This branch is 2 commits ahead of master.".
How can I put my `behst7` branch content in the master branch and propagate the new changes I did to the recipe?

Thanks

davidechicco commented:
I just realized I had to select "Squash and merge". Now it worked, thanks.
